### PR TITLE
[mm-604] - ability to create a subscription template

### PR DESCRIPTION
**Pull request by maisnamrajusingh**
#### Summary
A UI to create a subscription template. Please check the video below for details. 
https://www.loom.com/share/f988879135c54d789f4bd015b207e9d8
#### Ticket Link
https://github.com/mattermost/mattermost-plugin-jira/issues/604

### CODE REVIEW

File: server/subscribe.go [hunk 1]
```diff
@@ -936,6 +1064,77 @@ func (p *Plugin) httpChannelGetSubscriptions(w http.ResponseWriter, r *http.Requ
...
+	}
+
+	return respondJSON(w, subscriptionTemplates)
+}
+
+func (p *Plugin) httpCreateSubscriptionTemplate(w http.ResponseWriter, r *http.Request, mattermostUserID string) (int, error) {
+	subscriptionTemplate := SubscriptionTemplate{}
+	currentChannelID := r.FormValue("channel_id")
+
+	p.API.LogDebug(fmt.Sprintf("channel Id ------------------------------------------%s", currentChannelID))
```
Comment (sanjaydemansol) [inline server/subscribe.go:1088]: Forgot to remove LogDebug calls, will.

**Comment by mickmister** [comment]
@aaronrothschild @matthewbirtch Please see the video attached in the PR description. Currently it looks like the feature works this way:

- Opening the subscription modal shows the current channel's subscriptions and existing subscription templates
- The templates can be deleted from here
    - @maisnamrajusingh is there a confirmation to delete them after pressing `Delete`?
- When editing or creating a subscription, the form gives you the option to start from an existing template. When you select a template, it pre-populates the form with the template's values.
- Selecting "Save as Template" causes the current form state to be saved as a template.
    - @maisnamrajusingh What mechanisms are in place for overwriting a template, or saving a subscription with the same name as a template?
    - We should make sure the current user can access the template's Jira project on the server when processing the API request to save the template

@maisnamrajusingh Are the subscriptions scoped to a given channel or team? If possible, we should filter on Jira projects the current user has access to.

**Comment by matthewbirtch** [comment]
> @aaronrothschild @matthewbirtch Please see the video attached in the PR description

@mickmister do you know if any design work was done for this feature? I don't see anything done in our Figma files yet and I don't see a Figma link in the description here. I think the design team might need to take a pass at this to clean up the design/layout a bit.

**Comment by mickmister** [comment]
@matthewbirtch I don't think there's been any design for the feature, which was an oversight on creating the help wanted ticket. I'll discuss with the team how we can make sure design decisions are taken care of before opening the ticket for development.

What would be next steps to get this feature under design review?

**Comment by andrewbrown00** [comment]
> What would be next steps to get this feature under design review?

Hey, @mickmister, it would be helpful to know the use cases and requirements for this feature.

cc @aaronrothschild

**Comment by aaronrothschild** [comment]
I'll keep these use cases on this ticket:
- As a user, I want to create subscriptions to events in Jira so I can be notified of tickets I need to review/comment/approve/etc. directly into a Mattermost channel that is relevant, such as Project A.   (we have this functionality for admins)
- As an admin user I need to create subscriptions on behalf of users because they don't have permissions to create their own subscriptions, however - this becomes very repetitive when creating subscriptions for multiple users or channels because they need to re-enter the same values for each subscription - with a small variance.  
     - example: I need to subscribe to all bugs/stories/epics for project XYZ, but exclude all issues marked as "secure". I need to repeat this subscription for 10 new projects and channels.

Requirements:
- Only users permitted to create subscriptions in a channel can use this functionality
- Allow only Admin users to create a template that can be re-used when making the next subscription

     
     cc @matthewbirtch @mickmister

**Comment by mickmister** [comment]
@andrewbrown00 Our internal use case is that some of our Jira tickets are internal, and so we need to make sure all of the subscriptions going to public channels have these tickets filtered out. With this ticket implemented, we can have a baseline for a "safe" subscription to be reused among several users. This avoids user error that may expose sensitive information. We can then open up the feature to more users, since community currently has the feature locked to system admins.

The subscription template is identical to an actual subscription, except it doesn't actually process any webhook events, and it's not associated with any specific channel. It does need to adhere to every rule of subscriptions, meaning it needs to be scoped to a specific Jira project, and the user configuring the subscription needs to have access to that project in Jira.

**Comment by mickmister** [comment]
@matthewbirtch @andrewbrown00 I have two cloud servers set up for testing this. The first link is running the current release of the Jira plugin, and the second has this PR deployed. Access the feature by running `/jira subscribe` while signed into the usual "sysadmin" user

https://jira-plugin.test.mattermost.cloud
https://jira-plugin-pr-783.test.mattermost.cloud

**Comment by mickmister** [comment]
@matthewbirtch Thanks for all the detailed feedback! The channel subscriptions feature was developed a while ago and hasn't been revisited for some time. Reading through your comment, I agree that the overall flow of the feature can be improved, even outside of this PR's scope. The subscription template feature may need to put on hold if the redesign/reimplementation of the subscriptions feature needs to occur first, since that should be a separate PR from introducing the templates depending on how big the changes are.

**Comment by mickmister** [comment]
@matthewbirtch Would you like to have a video call to discuss this feature? If you'd like to test the feature yourself, please see the spinwick URLs posted here https://github.com/mattermost/mattermost-plugin-jira/pull/783#issuecomment-899842388

**Comment by matthewbirtch** [comment]
@mickmister I think that would likely be a good idea. Once we do that, we should probably this up for design to work through before developing further. Can we set something up with you, me @andrewbrown00 and @aaronrothschild?

**Comment by maisnamrajusingh** [comment]
@mickmister @matthewbirtch how did it go ?

**Comment by mickmister** [comment]
@andrewbrown00 @matthewbirtch @aaronrothschild

One thing that was not discussed is the access control of these templates. Depending on the structure of an organization, this may be a complicated equation. If we want to support the same levels of access control that we do elsewhere in the plugin, we will need to perform some heavy operations whenever someone chooses to browse the templates. Because of this, we should make the user go out of their way to see the list of available templates. i.e. It should be clear that the user wants to create a subscription from a template before we calculate which templates they have access to.

In fact, it would be helpful if the user is able to pick the template by project first. A subscription and subscription template must be tied to a specific Jira project, since all the facets of a subscription are tied to that project. If we allowed the user to first browse the subscription templates by project, that would be very useful. Jira projects that the user is not a member of would not show up in this list.

**Comment by aaronrothschild** [comment]
> @andrewbrown00 @matthewbirtch @aaronrothschild
> 
> One thing that was not discussed is the access control of these templates. Depending on the structure of an organization, this may be a complicated equation. If we want to support the same levels of access control that we do elsewhere in the plugin, we will need to perform some heavy operations whenever someone chooses to browse the templates. Because of this, we should make the user go out of their way to see the list of available templates. i.e. It should be clear that the user wants to create a subscription from a template before we calculate which templates they have access to.
> 
> In fact, it would be helpful if the user is able to pick the template by project first. A subscription and subscription template must be tied to a specific Jira project, since all the facets of a subscription are tied to that project. If we allowed the user to first browse the subscription templates by project, that would be very useful. Jira projects that the user is not a member of would not show up in this list.

Can you expand on the use case for security of the templates?  I think it would be fine for us to start with no access control for particular templates and perhaps add it later if requested/needed.  I understand that we'd want to follow your advice for now perhaps and show subscriptions based on Project to make it easier/possible in the future to have access control.

**Comment by hanzei** [comment]
@maisnamrajusingh Is this PR review for review?

**Comment by maisnamrajusingh** [comment]
> @maisnamrajusingh Is this PR review for review?

@hanzei no not yet. Currently prioritizing the jira ones.

**Comment by mickmister** [comment]
Closing in favor of https://github.com/mattermost/mattermost-plugin-jira/pull/897

**Pull request closed without merging**